threads/2024/THREADS-07-09.md

```diff
@@ -0,0 +1,205 @@
+
+TL: Time out. Conrad has offered to write up a fuller response when he has time.
+
+CW: Exceptions are complicated and we would need to figure out how to paper over the JS tag thing.
+
+AR: We need something more modular.
+
+CW: The other option would be to go whole hog for the effects, and you can alternatively view that as value passing. All control flow needs to handle it.
+
+LW: Compiler will not have issues because it already generates safe code.
+
+AR: But typing it is far too complicated.
+
+TL: Thread-local globals so far have had this peanut butter cost where they make all indirect calls slow because the TLS pointer might need to be updated on every potential instance boundary.
+
+RH: They would similarly make stack switches slow.
+
+TL: My latest thinking is that if we can have a dense vector of TLS blocks hanging off the instance, then maybe we can just do the lookup on each thread-local global access. It will only be a sequence of loads, not a concurrent hash table lookup.
+
+RH: Thought about this implementation. How do you allocate thread IDs? Do you reuse slots? With weak memory, reusing slots might require a barrier.
+
+TL: Good questions. I don't have answers yet. But I agree that paying the cost on every indirect call is bad.
+
+RH: For shadow stack pointer, I imagine it would still be faster to pass it around as an argument.
+
+TL: That would be good to experiment with.
+
+AR: How bad would it be if you just didn't do anything for TLS? Use ordinary parameter passing or effect handlers. There are several options.
+
+CW: That means we're deferring work stealing. We could do the version with shared non-suspendable and use non-shared parameters, but not more.
+
+RH: It could be useful to split the problem between shadow stack pointers and other use cases. Then we could still have shared-suspendable and pass the shadow stack pointer as a param.
+
+SYG: On performance and whether it's good enough to do something naive. It seems analogous to the security thing V8 is doing to sandbox pointers via a trusted external pointer table. It's not all pointers. I assume the TLS stuff won't be for all accesses. The sandbox is a concurrent, thread-safe pointer table that can grow and shrink based on GC pressure. The kind of pointers they want to sandbox on real-world workloads, the overall regression was 1-2%. We're willing to eat the cost for security. Is 1% an acceptable cost?
+
+CW: Even for shared-suspendable functions, we can have something that works. If you have thread-bound functions, the ABI could have you pass in all the thread-bound functions as shared parameters instead of unshared parameters.
+
+CW: We want to do shared functions. It's the new capability for the platform.
+
+TL: Shared structs and arrays are also a new capability for the platform, or at least get WasmGC languages to parity with linear memory languages.
+
+CW: There are three levels...
```

Comment:
@conrad-watt, do you remember what you were saying here?

Reply:
CW: there are three levels of complexity in supporting shared functions. (1) shared-nonsuspendable with only parameter passing, (2) shared-suspendable by adding thread-bound JS functions, (2) shared-suspendable by adding true Wasm-level TLS features

The point being that it would be good to at least do (1) rather than totally giving up on shared functions!